### PR TITLE
feat: CI Proxy upgrade verification flow

### PR DESCRIPTION
**Pull request by viraj124**
- closes https://github.com/FuelLabs/fuel-bridge/issues/205

### CODE REVIEW

File: packages/solidity-contracts/scripts/hardhat/verifyDeployment.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+          );
+          throw new Error('Init Code mismatch');
+        }
+
+        console.log(
+          '--- Check the new implementation deployment resulted in deploying that implementation addresss...'
+        );
+
+        if (
+          reciept?.contractAddress === deployment.linkedData.newImplementation
```
Comment (DefiCake) [inline packages/solidity-contracts/scripts/hardhat/verifyDeployment.ts:81]: Let's save the new implementation directly in the `implementation` field of the deployment artifact. We do not really need this `newImplementation` field in `linkeData`. Ideally the flow of a deployment is the following:
- We open a new PR. We propose the upgrade with `prepareUpgrade`, which also deploys the new implementation
- The multisig owners check the CI, and proceed to change the implementation of the proxy to the new proposed one.
- We merge the PR without having to modify any deployment artifact.

File: packages/solidity-contracts/scripts/hardhat/verifyDeployment.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+      if (!deployment.linkedData.isProxy) {
+        console.log('--- Creating local Hardhat network...');
+        const localHardhat = require('hardhat');
+        await localHardhat.run('compile');
+
+        const ContractFactory = await localHardhat.ethers.getContractFactory(
+          deployment.linkedData.factory
+        );
+
+        console.log('--- Validating Upgrade...');
```
Comment (DefiCake) [inline packages/solidity-contracts/scripts/hardhat/verifyDeployment.ts:30]: I think that either here or somewhere else where it makes sense, we should add a reference to the new implementation address that we are checking

**Review by DefiCake — COMMENTED** [review]
Looks sound. We need a cleanup and some polishing.

File: packages/solidity-contracts/scripts/hardhat/verifyDeployment.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+        const receipt = await ethers.provider.getTransactionReceipt(
+          fetchedDeploymentTx?.hash!
+        );
+
+        const txPayload = await hre.ethers.provider.getTransaction(
+          deployment.transactionHash!
+        );
+
+        if (expectedInitCode === txPayload?.data) {
```
Comment (DefiCake) [inline packages/solidity-contracts/scripts/hardhat/verifyDeployment.ts:66]: Add a check that `data` is not nullish

**Review by DefiCake — COMMENTED** [review]
lgtm, almost there, just a few nits

File: packages/solidity-contracts/scripts/hardhat/verifyDeployment.ts [hunk 4]
```diff
@@ -0,0 +1,123 @@
+import { task } from 'hardhat/config';
+import { HardhatRuntimeEnvironment } from 'hardhat/types';
+import { config as dotEnvConfig } from 'dotenv';
+import { ContractFactory } from 'ethers';
+import { writeFileSync } from 'fs';
+
+dotEnvConfig();
```
Comment (DefiCake) [inline packages/solidity-contracts/scripts/hardhat/verifyDeployment.ts:7]:
```suggestion
import { ContractFactory } from 'ethers';
import { writeFileSync } from 'fs';
```

Hardhat already does dotenv configuration, is it needed here?

File: .github/workflows/manual-verify-upgrade.yml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    inputs:
+      network:
+        description: 'Enter the network you want to verify the upgrade on'
+        required: true
+        default: 'mainnet'
+        type: string
+      rpc:
+        description: 'Enter network rpc'
+        required: true
+        type: string
```
Comment (DefiCake) [inline .github/workflows/manual-verify-upgrade.yml:14]:
```suggestion
      rpc:
        description: 'Enter network rpc'
        required: true
        default: 'https://eth.llamarpc.com'
        type: string
```

**Review by DefiCake — APPROVED** [review]
Left small suggestions, but lgtm

File: packages/solidity-contracts/scripts/hardhat/verifyDeployment.ts (hunk 4 above)
Comment (viraj124) [reply packages/solidity-contracts/scripts/hardhat/verifyDeployment.ts]: yeah removed

**Review by PraneshASP — APPROVED** [review]

**Review by DefiCake — APPROVED** [review]

**Pull request merged**